latest_major_version.go

```diff
@@ -0,0 +1,117 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"strings"
+	"os"
+	"regexp"
+	"encoding/json"
+	"net/http"
+	// "time"
+)
+
+func getLatestMajorVersion(repo string) (string, error) {
+	// Get latest major version available for repo from github.
+	const apiURL = "https://api.github.com/repos/%s/tags"
+	url := fmt.Sprintf(apiURL, repo)
+
+	resp, err := http.Get(url) 	// TODO: figure out ratelimiting??
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("failed to fetch tags: %s\n", resp.Status)
+	}
+
+	var tags []struct {
+		Name string `json:"name"`
+	}
+
+	if err := json.NewDecoder(resp.Body).Decode(&tags); err != nil {
+		return "", err
+	}
+
+	latest := ""
+	// match on version strings e.x. v9.3.0
+	majorVersionRegex := regexp.MustCompile(`^v(\d+)`) 
+	alphaBetaRegex := regexp.MustCompile(`alpha|beta`)
+
+
+	// Check latest major version seen
+	for _, tag := range tags {
+		if majorVersionRegex.MatchString(tag.Name) && !alphaBetaRegex.MatchString(tag.Name) {
+			version := majorVersionRegex.FindString(tag.Name)
+			if latest == "" || version > latest {
+				latest = version
+			}
+		}
+	}
+
+	return latest, nil
+}
+
+
+func main() {
+	file, err := os.Open("go.mod")
+	if err != nil {
+		fmt.Println("Error opening go.mod:", err)
+		return
+	}
+	defer file.Close()
+
+	// initialize map: module -> latest major version
+	latestMajor := make(map[string]string)
+
+	scanner := bufio.NewScanner(file)
+
+	// Match on versions with /v{major}
+	versionRegex := regexp.MustCompile(`^(?P<module>.+?)/v(\d+) v(\d+\.\d+\.\d+)$`)
```

Comment:
You can use https://pkg.go.dev/golang.org/x/mod/modfile to parse the `go.mod` file in an easier way.